### PR TITLE
Disallow the Cluster and Git resource types as Output resources.

### PR DESCRIPTION
**Pull request by dlorenc**
# Changes

The cluster type currently has no defined behavior as output resources, and
the git type behaves incorrectly. The new Volume type should be used instead
for the use-cases supported by the old Git resource.

This is part of the large cleanup in #1076, and should not be submitted until
after #1087.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

# Release Notes

```
The Git and Cluster PipelineResource types can no longer be used as an Output resource.
To replace the Git type, please use the Volume type. The Cluster type previously had no effect, so simply remove it as an Output.

```

### CODE REVIEW

**Comment by bobcatfish** [comment]
imo this counts as a backwards incompatible change, in which case I think we'd want to do something like:

1. warn in logs when ppl use git as an output
2. make sure we get the new volume type into this release
3. make it pretty clear in the working group that we're doing this
4. only actually remove in the next release (or even the one after)

I think (as you're seeing in the test failures) that it might be impossible to do input output linking in our tests with out this? (i think to use GCS we would need to surmount https://github.com/tektoncd/pipeline/issues/821 🤔🤔🤔

Also need to update:
* https://github.com/tektoncd/pipeline/blob/master/examples/pipelineruns/output-pipelinerun.yaml also
* https://github.com/tektoncd/pipeline/tree/master/docs/developers#how-are-outputs-handled
* Maybe add something at https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#git-resource (maybe we should make a table at the beginning or something showing which can be inputs and which can be outputs?)

(I'm definitely for doing this in the long run! I think this is gonna be a pretty disruptive change to anyone using linking tho since using a `git` output was pretty much the go-to way to do this)

**Comment by abayer** [comment]
I agree with @bobcatfish - Jenkins X's pipelines all utilize `git` as `output`, e.g., so a transition period is needed, even given that this states not to merge it until #1087 lands.

File: pkg/apis/pipeline/v1alpha1/resource_types.go [hunk 1]
```diff
@@ -130,15 +130,29 @@ type PipelineResourceList struct {
 	Items           []PipelineResource `json:"items"`
 }
 
-// ResourceFromType returns a PipelineResourceInterface from a PipelineResource's type.
-func ResourceFromType(r *PipelineResource) (PipelineResourceInterface, error) {
+// InputResourceFromType returns a PipelineResourceInterface from a PipelineResource's type.
+func InputResourceFromType(r *PipelineResource) (PipelineResourceInterface, error) {
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/resource_types.go]: would it be crazy in Go to define a top level like map of types to `New` functions? e.g. something like:

```go
PipelineResourceTypeGit: NewGitResource,
PipelineResourceTypeImage: NewImageResource,
```

If case statements are clearer, just ignore me

File: pkg/apis/pipeline/v1alpha1/resource_types.go [hunk 2]
```diff
@@ -130,15 +130,29 @@ type PipelineResourceList struct {
...
 	switch r.Spec.Type {
 	case PipelineResourceTypeGit:
 		return NewGitResource(r)
-	case PipelineResourceTypeImage:
-		return NewImageResource(r)
 	case PipelineResourceTypeCluster:
 		return NewClusterResource(r)
+	default:
+		return InputOutputResourceFromType(r)
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/resource_types.go]: i found it confusing that the default for this function calls `InputOutputResourceFromType` - id find it a bit easier to see explicitly which types are allowed for input and output, and always default to an error

**Comment by bobcatfish** [comment]
@sbwsg and @dlorenc have a new design to propose for #1272 and #1076 which I think will make this change unnecessary since they are proposing overhauling the way we express these resources completely, so I'm going to close this PR for now, but we can re-open if we decide we want to keep making these incremental improvements to PipelineResources as they are today (see also https://github.com/tektoncd/pipeline/issues/1062#issuecomment-546439474)

**Comment by bobcatfish** [comment]
p.s. I may open a PR to disallow cluster outputs anyway since that doesnt really make sense

**Pull request closed without merging**